tacacs_global changes for ios_xr

tphoney opened the pull request.

tphoney commented:
@chburket @mikewiebe need a review...

chburket commented on tests/beaker_tests/tacacs_global/tacacs_global_provider_defaults.rb:
```diff
@@ -144,15 +144,18 @@
-    on(agent, cmd_str) do
-      search_pattern_in_output(stdout, { 'enable' => 'false' },
-                               false, self, logger)
+    if platform != :ios_xr
```
Should the entire teststep be in the ios_xr ``if`` block, since it is a nop for XR?  Otherwise it looks like the teststep was executed.

chburket commented:
Other than a nit, :+1:

tphoney commented:
@chburket step is skipped if it is ios.

mikewiebe commented:
@tphoney Can you include XR and NXOS test results?

tphoney commented:
@mikewiebe my nexus keeps blowing up running anything to do with the tacacs_global
> 500 Error: Backend processing error

Can you please try running it

mikewiebe commented:
@tphoney I will see to it that someone on our team runs these tests and post the results here.

mikewiebe commented:
@tphoney I ran the test on my n9k but I see the following issue:
```
root@n9k-110#puppet agent -t
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Caching catalog for n9k-110.cisco.com
Info: Applying configuration version '1465480453'
Notice: /Stage[main]/Main/Node[default]/Tacacs_global[default]/enable: enable changed 'false' to 'true'
Notice: /Stage[main]/Main/Node[default]/Tacacs_global[default]/key: defined 'key' as '44444444'
Notice: /Stage[main]/Main/Node[default]/Tacacs_global[default]/key_format: defined 'key_format' as '7'
Notice: /Stage[main]/Main/Node[default]/Tacacs_global[default]/timeout: defined 'timeout' as '2'
Error: /Stage[main]/Main/Node[default]/Tacacs_global[default]: Could not evaluate: uninitialized constant UNSUPPORTED_PROPS
Notice: Applied catalog in 0.27 seconds
```

Here is the traceback:
```
root@n9k-110#puppet agent -t --trace
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Caching catalog for n9k-110.cisco.com
Info: Applying configuration version '1465480468'
Notice: /Stage[main]/Main/Node[default]/Tacacs_global[default]/enable: enable changed 'false' to 'true'
Notice: /Stage[main]/Main/Node[default]/Tacacs_global[default]/key: defined 'key' as '44444444'
Notice: /Stage[main]/Main/Node[default]/Tacacs_global[default]/key_format: defined 'key_format' as '7'
Notice: /Stage[main]/Main/Node[default]/Tacacs_global[default]/timeout: defined 'timeout' as '2'
Error: /Stage[main]/Main/Node[default]/Tacacs_global[default]: Could not evaluate: uninitialized constant UNSUPPORTED_PROPS
/opt/puppetlabs/puppet/cache/lib/puppet/provider/tacacs_global/cisco.rb:94:in `validate'
/opt/puppetlabs/puppet/cache/lib/puppet/provider/tacacs_global/cisco.rb:121:in `flush'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type.rb:1004:in `flush'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction/resource_harness.rb:24:in `evaluate'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:212:in `apply'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:228:in `eval_resource'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:151:in `call'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:151:in `block (2 levels) in evaluate'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:386:in `block in thinmark'
/opt/puppetlabs/puppet/lib/ruby/2.1.0/benchmark.rb:294:in `realtime'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:385:in `thinmark'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:151:in `block in evaluate'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/graph/relationship_graph.rb:118:in `traverse'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:142:in `evaluate'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/catalog.rb:222:in `block in apply'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/log.rb:155:in `with_destination'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction/report.rb:118:in `as_logging_destination'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/catalog.rb:221:in `apply'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:171:in `block in apply_catalog'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:223:in `block in benchmark'
/opt/puppetlabs/puppet/lib/ruby/2.1.0/benchmark.rb:294:in `realtime'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:222:in `benchmark'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:170:in `apply_catalog'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:315:in `run_internal'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:186:in `block in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/context.rb:65:in `override'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet.rb:240:in `override'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:185:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:45:in `block (4 levels) in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent/locker.rb:21:in `lock'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:45:in `block (3 levels) in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:98:in `with_client'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:42:in `block (2 levels) in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:65:in `run_in_fork'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:41:in `block in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:179:in `call'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:179:in `controlled_run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/agent.rb:39:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/agent.rb:353:in `onetime'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/agent.rb:331:in `run_command'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:344:in `block in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:540:in `exit_on_fail'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:344:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:128:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:72:in `execute'
/opt/puppetlabs/puppet/bin/puppet:5:in `<main>'
Notice: Applied catalog in 0.27 seconds
root@n9k-110#
```

```

tphoney commented:
@mikewiebe can you try testing again, please.

tphoney commented:
@mikewiebe any luck re-running this ?

tphoney commented:
@mikewiebe @chburket based on the feedback and errors, i implemented tacacs_global in the same way as radius_global. There is now a tacacs_global type/provider in nodeutils everything. It is this PR https://github.com/cisco/cisco-network-node-utils/pull/429 

here is the passing test:
     Test Suite: tests @ 2016-06-15 14:55:58 +0100

      - Host Configuration Summary -


              - Test Case Summary for suite 'tests' -
       Total Suite Time: 116.81 seconds
      Average Test Time: 116.81 seconds
              Attempted: 1
                 Passed: 1
                 Failed: 0
                Errored: 0
                Skipped: 0
                Pending: 0
                  Total: 1

      - Specific Test Case Status -

Failed Tests Cases:
Errored Tests Cases:
Skipped Tests Cases:
Pending Tests Cases:

tphoney commented:
@chburket @mikewiebe now that cisco/cisco-network-node-utils#429 has been merged can you look at this PR.

chburket commented on tests/beaker_tests/tacacs_global/tacacs_global_provider_defaults.rb:
```diff
@@ -63,9 +62,7 @@
-    on(master, TacacsLib.create_tacacs_manifest_change_disabled)
-    cmd_str = PUPPET_BINPATH + 'agent -t'
-    on(agent, cmd_str, acceptable_exit_codes: [0, 2])
+    logger.info('Setup switch for provider')
```
This step no longer does anything?

chburket commented on tests/beaker_tests/tacacs_global/tacacs_global_provider_defaults.rb:
```diff
@@ -75,7 +72,7 @@
     # Expected exit_code is 2 since this is a puppet agent cmd with change.
     cmd_str = PUPPET_BINPATH + 'agent -t'
-    on(agent, cmd_str, acceptable_exit_codes: [2])
+    on(agent, cmd_str, acceptable_exit_codes: [0, 2])
```
Will this ever make a change, now?  If not, is it needed anymore?  No matter what, the comment is no longer valid.

tphoney commented:
@chburket the tests were re-written to more closely mimic what is currently running for the tests running against radius_global. The test layout has changed subtly.

chburket commented:
@tphoney That's fine, but my questions still seem valid:

1. You have a test step with no real code in it.  Is that intentional / ok?
2. You have a test that now accepts a return code of 0 as well as 2.  The comment says that only 2 is expected.  Something needs to change.

tphoney commented:
@chburket 
You have a test step with no real code in it. Is that intentional / ok? **now removed, this was mimicking radius_global_provider_defaults**
You have a test that now accepts a return code of 0 as well as 2. The comment says that only 2 is expected. Something needs to change. **I have updated the comments, the return value is dependant on the initial state of the switch, thus can vary**

chburket commented:
:+1:

tphoney commented:
@mikewiebe can i get a merge

mikewiebe commented:
@tphoney Were you able to re-run the beaker tests on an `n9kv`?  Last time I ran it I got an error and want to make sure we run clean now.

tphoney commented:
@mikewiebe this is re-written to mimic the radius_global provider, my nexus box is borked. Can you try this for me please.

tphoney commented:
@mikewiebe any luck ?

mikewiebe commented:
@tphoney No, not yet.  My testbed is currently tied up with other testing but I will try and execute by COB today.  If all is well we can merge.

mikewiebe commented:
@tphoney Ran the test but still see the following error.  I node_utils gem was latest develop.
```
2016-06-24T16:05:19: %SCRIPT-INFO:     	Info: Loading facts
2016-06-24T16:05:19: %SCRIPT-INFO:     	Info: Caching catalog for dt-n9k5-1.cisco.com
2016-06-24T16:05:19: %SCRIPT-INFO:     	Info: Applying configuration version '1466798718'
2016-06-24T16:05:19: %SCRIPT-INFO:     	Error: Failed to apply catalog: undefined method `default_key' for Cisco::TacacsGlobal:Class>
```

tphoney commented:
@chrisvanheuveln @mikewiebe more changes required, https://github.com/cisco/cisco-network-node-utils/pull/437 with the fixes we get the following passes. It took a day to set up the nexus device working again.
**nexus test output** 
ests/beaker_tests/tacacs_global/tacacs_global_provider_defaults.rb passed in 100.50 seconds
      Test Suite: tests @ 2016-06-27 15:31:49 +0100

      - Host Configuration Summary -


              - Test Case Summary for suite 'tests' -
       Total Suite Time: 100.50 seconds
      Average Test Time: 100.50 seconds
              Attempted: 1
                 Passed: 1
                 Failed: 0
                Errored: 0
                Skipped: 0
                Pending: 0
                  Total: 1

      - Specific Test Case Status -

Failed Tests Cases:
Errored Tests Cases:
Skipped Tests Cases:
Pending Tests Cases:
**IOS device**
tests/beaker_tests/tacacs_global/tacacs_global_provider_defaults.rb passed in 31.20 seconds
      Test Suite: tests @ 2016-06-27 15:35:24 +0100

      - Host Configuration Summary -


              - Test Case Summary for suite 'tests' -
       Total Suite Time: 31.20 seconds
      Average Test Time: 31.20 seconds
              Attempted: 1
                 Passed: 1
                 Failed: 0
                Errored: 0
                Skipped: 0
                Pending: 0
                  Total: 1

      - Specific Test Case Status -

Failed Tests Cases:
Errored Tests Cases:
Skipped Tests Cases:
Pending Tests Cases:

mikewiebe commented:
@tphoney Thanks for verifying.  I made some comments on cisco/cisco-network-node-utils#437.  Once they are addressed I will re-validate both minitest and beaker tests on nexus.
